File: patterns/2020_10_07/cppImpl/survey.cpp
#include <time.h>
#include <cstdio>
#include <ctime>
#include "nngpp/nngpp.h"
#include "nngpp/protocol/survey0.h"
#include "nngpp/protocol/respond0.h"

#define SERVER "server"
#define CLIENT "client"
#define DATE "DATE"

char *date(void)
{
    time_t now = time(&now);
    struct tm *info = localtime(&now);
    char *text = asctime(info);
    text[strlen(text) - 1] = '\0'; // remove '\n'
    return (text);
}

using namespace nng;
int server(const char *url)
{
    try
    {
        auto sock = survey::open();
        survey::set_opt_survey_time(sock, 1000);
        sock.listen(url);

        while (true)
        {
            printf("SERVER: SENDING DATE SURVEY REQUEST\n");
            sock.send("DATE");

            while (true)
            {
                try
                {
                    auto msg = sock.recv();
                    printf("SERVER: RECEIVED \"%s\" SURVEY RESPONSE\n", msg.data<char>());
                }
                catch (const nng::exception &e)
                {
                    if (e.get_error() == nng::error::timedout)
                        break;
                    else
                    {
                        printf("%s: %s\n", e.who(), e.what());
                        return 1;
                    }
                }
            }
            printf("SERVER:SURVERY COMPLETE\n");
        }
    }
    catch (const nng::exception &e)
    {
        printf("%s: %s\n", e.who(), e.what());
        return 1;
    }
}

int client(const char *url, const char *name)
{
    try
    {
        auto sock = respond::open();
        sock.dial(url);
        while (true)
        {
            auto msg = sock.recv();
            printf("CLIENT (%s): RECEIVED \"%s\" SURVERY REQUEST\n", name, msg.data<char>());
            char *now = date();
            printf("CLIENT (%s): SENDING DATE SURVEY RESPONSE\n", name);
            sock.send(view{now, strlen(now) + 1});
        }
    }
    catch (const nng::exception &e)
    {
        printf("%s: %s\n", e.who(), e.what());
        return 1;
    }
}

int main(const int argc, const char **argv)
{
    if ((argc >= 2) && (strcmp(SERVER, argv[1]) == 0))
        return (server(argv[2]));

    if ((argc >= 3) && (strcmp(CLIENT, argv[1]) == 0))
        return (client(argv[2], argv[3]));

    fprintf(stderr, "Usage: surveyImpl %s|%s <URL> <ARG>...\n", SERVER, CLIENT);
    return (1);
}

File: patterns/2020_10_07/pythonImpl/survey.py
import sys
import time
from datetime import datetime
from pynng import Surveyor0, Respondent0, Timeout


def server(url: str):
    with Surveyor0(survey_time=100, listen=url) as sock:
        while True:
            print(f'SERVER: SENDING DATE SURVEY REQUEST')
            sock.send('DATE'.encode())

            while True:
                try:
                    msg = sock.recv()
                    print(f'SERVER: RECEIVED "{msg.decode()}" SURVEY RESPONSE')
                except Timeout:
                    break

            print(f'SERVER: SURVEY COMPLETE')


def client(url: str, name: str):
    with Respondent0(dial=url, block_on_dial=False, recv_timeout=100) as sock:
        while True:
            try:
                msg = sock.recv()

                if str(msg.decode()) == 'DATE':
                    print(
                        f'CLIENT ({name}): RECEIVED "{msg.decode()}" SURVEY REQUEST')
                    print(f'CLIENT ({name}): SENDING DATE SURVEY RESPONSE')
                    data = str(datetime.now())
                    sock.send(data.encode())
            except Timeout:
                pass


if __name__ == "__main__":
    # print(sys.argv)
    if len(sys.argv) > 2 and sys.argv[1] == 'server':
        server(sys.argv[2])
    elif len(sys.argv) > 3 and sys.argv[1] == 'client':
        client(sys.argv[2], sys.argv[3])
    else:
        print(f"Usage: {sys.argv[0]} server|client <URL> <ARGS> ...")
        sys.exit(1)
